### PR TITLE
Better diagnostics for faulty installations.

### PR DESCRIPTION
**Pull request by Volker-Weissmann**
The whole cygdb construct is quite messy and a faulty installation is hard to debug because the error messages don't really hint at the problem.
This PR slightly improves the situation.

### CODE REVIEW

File: Cython/Debugger/Cygdb.py [hunk 1]
```diff
@@ -45,17 +45,22 @@ def make_command_file(path_to_debug_info, prefix_code='', no_import=False):
...
+                    path_to_activate_this_py = os.path.join(virtualenv, 'bin', 'activate_this.py')
+                    print("gdb command file: Activating virtualenv: %s; path_to_activate_this_py: %s" % (
+                        virtualenv, path_to_activate_this_py))
+                    with open(path_to_activate_this_py) as f:
+                        exec(f.read(), dict(__file__=path_to_activate_this_py))
+                from Cython.Debugger import libcython, libpython
+            except Exception as ex:
+                print("There was an error in Python code originating from the file ''' + str(__file__) + '''")
+                print("It used the Python interpreter " + str(sys.executable))
+                print_exc()
```
Comment (scoder) [inline Cython/Debugger/Cygdb.py:63]: Why not re-raise the exception here?

File: Cython/Debugger/Cygdb.py [hunk 2]
```diff
@@ -45,17 +45,22 @@ def make_command_file(path_to_debug_info, prefix_code='', no_import=False):
...
+                virtualenv = os.getenv('VIRTUAL_ENV')
+                if virtualenv:
+                    path_to_activate_this_py = os.path.join(virtualenv, 'bin', 'activate_this.py')
+                    print("gdb command file: Activating virtualenv: %s; path_to_activate_this_py: %s" % (
+                        virtualenv, path_to_activate_this_py))
+                    with open(path_to_activate_this_py) as f:
+                        exec(f.read(), dict(__file__=path_to_activate_this_py))
+                from Cython.Debugger import libcython, libpython
+            except Exception as ex:
+                print("There was an error in Python code originating from the file ''' + str(__file__) + '''")
```
Comment (scoder) [inline Cython/Debugger/Cygdb.py:61]: Is `__file__` really the right file to report here? Or should it point at the virtualenv (if any) that it just tried to initialise?

File: Cython/Debugger/Cygdb.py (hunk 1 above)
Comment (Volker-Weissmann) [reply Cython/Debugger/Cygdb.py:63]: print_exc gives a better error message than raising:
Example: I added a raise ValueErrror("test") to libpython.py
Without re-raising we get:

> There was an error in Python code originating from the file /home/volker/Sync/git/cython/Cython/Debugger/Cygdb.py
> It used the Python interpreter /usr/bin/python
> Traceback (most recent call last):
>   File "<string>", line 12, in <module>
>   File "/home/volker/Sync/git/cython/Cython/Debugger/libcython.py", line 60, in <module>
>     from Cython.Debugger import libpython
>   File "/home/volker/Sync/git/cython/Cython/Debugger/libpython.py", line 57, in <module>
>     raise ValueError("test")
> ValueError: test
> /tmp/tmpv_wmeqcm:28: Error in sourced command file:
> Undefined command: "cy".  Try "help".
> (gdb) 
With re-raising and without print_exc() we get:

> There was an error in Python code originating from the file /home/volker/Sync/git/cython/Cython/Debugger/Cygdb.py
> It used the Python interpreter /usr/bin/python
> Traceback (most recent call last):
>   File "<string>", line 18, in <module>
> ValueError: test
> /tmp/tmphlesqyrq:26: Error in sourced command file:
> Error while executing Python code.
> (gdb) 
But I just realised that exit(1) will stop gdb wich is what we want. I added a commit.

File: Cython/Debugger/Cygdb.py (hunk 2 above)
Comment (Volker-Weissmann) [reply Cython/Debugger/Cygdb.py:61]: I think that __file__ is correct I installed it in a virtualenv and it printed:
/home/volker/Documents/venv/lib/python3.8/site-packages/Cython/Debugger/Cygdb.py
which is exactly what we want.

File: Cython/Debugger/Cygdb.py [hunk 3]
```diff
@@ -85,8 +85,8 @@ def make_command_file(path_to_debug_info, prefix_code='', no_import=False):
                     gdb.lookup_type('PyModuleObject')
                 except RuntimeError:
                     sys.stderr.write(
-                        'Python was not compiled with debug symbols (or it was '
-                        'stripped). Some functionality may not work (properly).\\n')
+                        "''' + interpreter + ''' was not compiled with debug symbols (or it was "
```
Comment (scoder) [inline Cython/Debugger/Cygdb.py:88]: Did you actually try this? I don't think it inserts the interpreter name here.

File: Cython/Debugger/Cygdb.py [hunk 4]
```diff
@@ -45,17 +45,23 @@ def make_command_file(path_to_debug_info, prefix_code='', no_import=False):
...
-            virtualenv = os.getenv('VIRTUAL_ENV')
-            if virtualenv:
-                path_to_activate_this_py = os.path.join(virtualenv, 'bin', 'activate_this.py')
-                print("gdb command file: Activating virtualenv: %s; path_to_activate_this_py: %s" % (
-                    virtualenv, path_to_activate_this_py))
-                with open(path_to_activate_this_py) as f:
-                    exec(f.read(), dict(__file__=path_to_activate_this_py))
-
-            from Cython.Debugger import libcython, libpython
+            from traceback import print_exc
```
Comment (scoder) [inline Cython/Debugger/Cygdb.py:48]: I think this should only get imported in the exception case.

Comment (Volker-Weissmann) [reply Cython/Debugger/Cygdb.py]: Ok, I moved it.

File: Cython/Debugger/Cygdb.py [hunk 5]
```diff
@@ -79,8 +85,8 @@ def make_command_file(path_to_debug_info, prefix_code='', no_import=False):
                     gdb.lookup_type('PyModuleObject')
                 except RuntimeError:
                     sys.stderr.write(
-                        'Python was not compiled with debug symbols (or it was '
-                        'stripped). Some functionality may not work (properly).\\n')
+                        "''' + interpreter + ''' was not compiled with debug symbols (or it was "
```
Comment (Volker-Weissmann) [inline Cython/Debugger/Cygdb.py:88]: Someone commented "Did you actually try this? I don't think it inserts the interpreter name here.", but I can't find find the comment.
Yes I tried this. interpreter is the variable defined by interpreter = interpreter_file.read() . Note that this is not the current python interpreter (that would be sys.executable) but the interpreter that compiled the users .pyx files.
Example:
```
$ python2 setup.py build_ext --inplace   
...
$ cygdb . -- -q --args python main.py
Reading symbols from python...
(No debugging symbols found in python)
/usr/bin/python2 was not compiled with debug symbols (or it was stripped). Some functionality may not work (properly).
(gdb)
```

Comment (scoder) [reply Cython/Debugger/Cygdb.py:88]: Yeah, my bad. I failed to understand the string quoting in the diff. All fine.

**Comment by scoder** [comment]
Thanks

**Pull request merged**